src/Agent.Service/Windows/AgentListener.cs

```diff
@@ -0,0 +1,276 @@
+                }
+
+                if(_listenerProcess != null)
+                {
+                    ProcessHelper.StopProcess(_listenerProcess.Id);
+                }
+            }
+        }
+
+        private void CreateAndStartAgentListenerProcess()
+        {
+            _listenerProcess = CreateAgentListener();
+            _listenerProcess.OutputDataReceived += AgentListener_OutputDataReceived;
+            _listenerProcess.ErrorDataReceived += AgentListener_ErrorDataReceived;
+            _listenerProcess.Start();
+            _listenerProcess.BeginOutputReadLine();
+            _listenerProcess.BeginErrorReadLine();
+        }
+
+        private void DisposeAgentListenerProcess()
+        {
+            _listenerProcess.OutputDataReceived -= AgentListener_OutputDataReceived;
+            _listenerProcess.ErrorDataReceived -= AgentListener_ErrorDataReceived;
+            _listenerProcess.Dispose();
+            _listenerProcess = null;
+        }
+
+        private int HandleExitOfListenerProcess(int exitCode)
+        {
+            int finalExitCode = exitCode;
+
+            // exit code 0 and 1 need stop service
+            // exit code 2 and 3 need restart agent
+            switch (exitCode)
+            {
+                case 0:
+                    Stopping = true;
+                    EventLogger.WriteInfo(Resource.AgentExitWithoutError);
+                    break;
+                case 1:
+                    Stopping = true;
+                    EventLogger.WriteInfo(Resource.AgentExitWithTerminatedError);
+                    break;
+                case 2:
+                    EventLogger.WriteInfo(Resource.AgentExitWithError);
+                    break;
+                case 3:
+                    EventLogger.WriteInfo(Resource.AgentUpdateInProcess);
+                    var updateResult = HandleAgentUpdate();
+                    if (updateResult == AgentUpdateResult.Succeed)
+                    {
+                        EventLogger.WriteInfo(Resource.AgentUpdateSucceed);
+                    }
+                    else if (updateResult == AgentUpdateResult.Failed)
+                    {
+                        EventLogger.WriteInfo(Resource.AgentUpdateFailed);
+                        Stopping = true;
+                    }
+                    else if (updateResult == AgentUpdateResult.SucceedNeedRestart)
+                    {
+                        EventLogger.WriteInfo(Resource.AgentUpdateRestartNeeded);
+                        _restart = true;                                
+                        exitCode = int.MaxValue;
+                        Stop();
+                    }
+                    break;
+                default:
+                    EventLogger.WriteInfo(Resource.AgentExitWithUndefinedReturnCode);
+                    break;
+            }
+
+            return finalExitCode;
+        }        
+
+        private AgentUpdateResult HandleAgentUpdate()
```

Comment:
I think in order for agent update to run smoothly, you will need to ensure the agent service process is either A) started with the bin/.. dir as the working directory of the process, or B) after the process starts up, cd ..